File: cpp/build-company/ConstructionCompany.cpp
#include "ConstructionCompany.hpp"
#include <iostream>

// Реализация конструктора
ConstructionCompany::ConstructionCompany(const std::string& companyName)
    : companyName(companyName) {
    std::cout << "Создана строительная компания '" << companyName << "'" << std::endl;
}

// Реализация деструктора
ConstructionCompany::~ConstructionCompany() {
    std::cout << "Вызван деструктор ConstructionCompany для: " << companyName << std::endl;
}

// Реализация геттеров
std::string ConstructionCompany::getCompanyName() const {
    return companyName;
}

const std::vector<Staff*>& ConstructionCompany::getStaff() const {
    return staff;
}

const std::vector<Warehouse*>& ConstructionCompany::getWarehouses() const {
    return warehouses;
}

const std::vector<ConstructionObject*>& ConstructionCompany::getConstructionObjects() const {
    return constructionObjects;
}

// Реализация сеттера
void ConstructionCompany::setCompanyName(const std::string& companyName) {
    this->companyName = companyName;
}

// Добавление сотрудника
void ConstructionCompany::addStaff(Staff* person) {
    if (person != nullptr) {
        staff.push_back(person);
        std::cout << "Добавлен сотрудник: " << person->getFullname() << std::endl;
    }
}

// Удаление сотрудника по индексу
void ConstructionCompany::deleteStaff(int staffId) {
    if (staffId >= 0 && staffId < static_cast<int>(staff.size())) {
        std::cout << "Удален сотрудник: " << staff[staffId]->getFullname() << std::endl;
        staff.erase(staff.begin() + staffId);
    } else {
        std::cout << "Ошибка: Неверный ID сотрудника" << std::endl;
    }
}

// Добавление склада
void ConstructionCompany::addWarehouse(Warehouse* warehouse) {
    if (warehouse != nullptr) {
        warehouses.push_back(warehouse);
        std::cout << "Добавлен склад по адресу: " << warehouse->getAddress() << std::endl;
    }
}

// Удаление склада по индексу
void ConstructionCompany::deleteWarehouse(int warehouseId) {
    if (warehouseId >= 0 && warehouseId < static_cast<int>(warehouses.size())) {
        std::cout << "Удален склад по адресу: " << warehouses[warehouseId]->getAddress() << std::endl;
        warehouses.erase(warehouses.begin() + warehouseId);
    } else {
        std::cout << "Ошибка: Неверный ID склада" << std::endl;
    }
}

// Добавление строительного объекта
void ConstructionCompany::buildNewObject(ConstructionObject* object) {
    if (object != nullptr) {
        constructionObjects.push_back(object);
        std::cout << "Начат новый строительный проект по адресу: " << object->getAddress() << std::endl;
    }
}

// Удаление строительного объекта по индексу
void ConstructionCompany::sellObject(int objectId) {
    if (objectId >= 0 && objectId < static_cast<int>(constructionObjects.size())) {
        std::cout << "Продан/Завершен проект по адресу: " << constructionObjects[objectId]->getAddress() << std::endl;
        constructionObjects.erase(constructionObjects.begin() + objectId);
    } else {
        std::cout << "Ошибка: Неверный ID объекта" << std::endl;
    }
}

// Отображение информации о компании
void ConstructionCompany::displayInfo() const {
    std::cout << "\n========================================" << std::endl;
    std::cout << "=== Информация о строительной компании ===" << std::endl;
    std::cout << "========================================" << std::endl;
    std::cout << "Название компании: " << companyName << std::endl;
    std::cout << "Всего сотрудников: " << staff.size() << std::endl;
    std::cout << "Всего складов: " << warehouses.size() << std::endl;
    std::cout << "Всего строительных проектов: " << constructionObjects.size() << std::endl;
    std::cout << "========================================\n" << std::endl;
}

// Отображение всех сотрудников
void ConstructionCompany::displayAllStaff() const {
    std::cout << "\n=== Все сотрудники ===" << std::endl;
    if (staff.empty()) {
        std::cout << "Сотрудники не зарегистрированы" << std::endl;
    } else {
        for (size_t i = 0; i < staff.size(); ++i) {
            std::cout << "Сотрудник #" << i << ":" << std::endl;
            staff[i]->displayInfo();
            std::cout << std::endl;
        }
    }
}

// Отображение всех складов
void ConstructionCompany::displayAllWarehouses() const {
    std::cout << "\n=== Все склады ===" << std::endl;
    if (warehouses.empty()) {
        std::cout << "Склады не зарегистрированы" << std::endl;
    } else {
        for (size_t i = 0; i < warehouses.size(); ++i) {
            std::cout << "Склад #" << i << ":" << std::endl;
            warehouses[i]->displayInfo();
            std::cout << std::endl;
        }
    }
}

// Отображение всех строительных объектов
void ConstructionCompany::displayAllObjects() const {
    std::cout << "\n=== Все строительные проекты ===" << std::endl;
    if (constructionObjects.empty()) {
        std::cout << "Строительные проекты не зарегистрированы" << std::endl;
    } else {
        for (size_t i = 0; i < constructionObjects.size(); ++i) {
            std::cout << "Проект #" << i << ":" << std::endl;
            constructionObjects[i]->displayInfo();
            std::cout << std::endl;
        }
    }
}

File: csharp/build-company/ConstructionCompany.cs
namespace BuildCompany;

/// <summary>
/// Главный класс, представляющий строительную компанию
/// </summary>
public class ConstructionCompany
{
    private string companyName;
    private List<Staff> staff;
    private List<Warehouse> warehouses;
    private List<ConstructionObject> constructionObjects;

    /// <summary>
    /// Конструктор
    /// </summary>
    /// <param name="companyName">Название компании</param>
    public ConstructionCompany(string companyName)
    {
        this.companyName = companyName;
        this.staff = new List<Staff>();
        this.warehouses = new List<Warehouse>();
        this.constructionObjects = new List<ConstructionObject>();
    }

    /// <summary>
    /// Получить название компании
    /// </summary>
    public string GetCompanyName()
    {
        return companyName;
    }

    /// <summary>
    /// Получить список персонала
    /// </summary>
    public List<Staff> GetStaff()
    {
        return staff;
    }

    /// <summary>
    /// Получить список складов
    /// </summary>
    public List<Warehouse> GetWarehouses()
    {
        return warehouses;
    }

    /// <summary>
    /// Получить список строительных объектов
    /// </summary>
    public List<ConstructionObject> GetConstructionObjects()
    {
        return constructionObjects;
    }

    /// <summary>
    /// Установить название компании
    /// </summary>
    public void SetCompanyName(string companyName)
    {
        this.companyName = companyName;
    }

    /// <summary>
    /// Добавить сотрудника
    /// </summary>
    public void AddStaff(Staff person)
    {
        staff.Add(person);
        Console.WriteLine($"Staff member '{person.GetFullname()}' added to {companyName}");
    }

    /// <summary>
    /// Удалить сотрудника
    /// </summary>
    public void DeleteStaff(int staffId)
    {
        if (staffId >= 0 && staffId < staff.Count)
        {
            var person = staff[staffId];
            staff.RemoveAt(staffId);
            Console.WriteLine($"Staff member '{person.GetFullname()}' removed from {companyName}");
        }
    }

    /// <summary>
    /// Добавить склад
    /// </summary>
    public void AddWarehouse(Warehouse warehouse)
    {
        warehouses.Add(warehouse);
        Console.WriteLine($"Warehouse at '{warehouse.GetAddress()}' added to {companyName}");
    }

    /// <summary>
    /// Удалить склад
    /// </summary>
    public void DeleteWarehouse(int warehouseId)
    {
        if (warehouseId >= 0 && warehouseId < warehouses.Count)
        {
            var warehouse = warehouses[warehouseId];
            warehouses.RemoveAt(warehouseId);
            Console.WriteLine($"Warehouse at '{warehouse.GetAddress()}' removed from {companyName}");
        }
    }

    /// <summary>
    /// Начать строительство нового объекта
    /// </summary>
    public void BuildNewObject(ConstructionObject obj)
    {
        constructionObjects.Add(obj);
        Console.WriteLine($"New construction project at '{obj.GetAddress()}' added to {companyName}");
    }

    /// <summary>
    /// Продать объект
    /// </summary>
    public void SellObject(int objectId)
    {
        if (objectId >= 0 && objectId < constructionObjects.Count)
        {
            var obj = constructionObjects[objectId];
            constructionObjects.RemoveAt(objectId);
            Console.WriteLine($"Construction project at '{obj.GetAddress()}' sold and removed from {companyName}");
        }
    }

    /// <summary>
    /// Отображение информации о компании
    /// </summary>
    public void DisplayInfo()
    {
        Console.WriteLine($"\n=== {companyName} ===");
        Console.WriteLine($"Total Staff: {staff.Count}");
        Console.WriteLine($"Total Warehouses: {warehouses.Count}");
        Console.WriteLine($"Total Projects: {constructionObjects.Count}");
    }

    /// <summary>
    /// Отображение всего персонала
    /// </summary>
    public void DisplayAllStaff()
    {
        Console.WriteLine($"\n=== All Staff Members of {companyName} ===");
        for (int i = 0; i < staff.Count; i++)
        {
            Console.WriteLine($"\nStaff [{i}]:");
            staff[i].DisplayInfo();
        }
    }

    /// <summary>
    /// Отображение всех складов
    /// </summary>
    public void DisplayAllWarehouses()
    {
        Console.WriteLine($"\n=== All Warehouses of {companyName} ===");
        for (int i = 0; i < warehouses.Count; i++)
        {
            Console.WriteLine($"\nWarehouse [{i}]:");
            warehouses[i].DisplayInfo();
        }
    }

    /// <summary>
    /// Отображение всех строительных объектов
    /// </summary>
    public void DisplayAllObjects()
    {
        Console.WriteLine($"\n=== All Construction Projects of {companyName} ===");
        for (int i = 0; i < constructionObjects.Count; i++)
        {
            Console.WriteLine($"\nProject [{i}]:");
            constructionObjects[i].DisplayInfo();
        }
    }
}
